File: android/mobile/src/main/java/com/novoda/peepz/SignInActivity.java
package com.novoda.peepz;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;

import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInResult;
import com.google.firebase.auth.FirebaseUser;

import butterknife.BindView;
import butterknife.ButterKnife;

public class SignInActivity extends BaseActivity {

    private static final int REQUEST_CODE_SIGN_IN = 1;

    @BindView(R.id.sign_in_button)
    Button signInButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (firebaseApi().isSignedIn()) {
            navigateToThingyAndFinish();
        } else {
            setContentView(R.layout.activity_sign_in);
            ButterKnife.bind(this);

            signInButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Intent intent = googleApiClientApi().getSignInIntent();
                    startActivityForResult(intent, REQUEST_CODE_SIGN_IN);
                }
            });
        }
    }

    private void navigateToThingyAndFinish() {
        startActivity(new Intent(this, PeepzActivity.class));
        finish();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_CODE_SIGN_IN) {
            GoogleSignInResult result = googleApiClientApi().getSignInResultFromIntent(data);
            if (result.isSuccess()) {
                signIntoFirebase(result);
            } else {
                displayError();
            }
        }
    }

    private void signIntoFirebase(GoogleSignInResult result) {
        GoogleSignInAccount googleSignInAccount = result.getSignInAccount();
        firebaseApi().signIntoFirebase(googleSignInAccount, new AuthenticationCallbacks() {
            @Override
            public void onSuccess(FirebaseUser firebaseUser) {
                navigateToThingyAndFinish();
            }

            @Override
            public void onAuthenticationFailure() {
                displayError();
            }
        });
    }

    private void displayError() {
        // TODO: show an error state, ideally with a reason
    }

}

File: android/mobile/src/main/java/com/novoda/peepz/SignInActivity.kt
package com.novoda.peepz

import android.content.Intent
import android.os.Bundle
import butterknife.ButterKnife
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.firebase.auth.FirebaseUser
import com.novoda.peepz.databinding.ActivitySignInBinding

class SignInActivity : BaseActivity() {
    private lateinit var binding: ActivitySignInBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (firebaseApi().isSignedIn) {
            navigateToThingyAndFinish()
        } else {
            binding = ActivitySignInBinding.inflate(layoutInflater)
            setContentView(binding.root)
            ButterKnife.bind(this)
            binding.signInButton.setOnClickListener {
                val intent = googleApiClientApi().signInIntent
                startActivityForResult(intent, REQUEST_CODE_SIGN_IN)
            }
        }
    }

    private fun navigateToThingyAndFinish() {
        startActivity(Intent(this, PeepzActivity::class.java))
        finish()
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_CODE_SIGN_IN) {
            val result = googleApiClientApi().getSignInResultFromIntent(data)
            if (result.isSuccess) {
                signIntoFirebase(result.signInAccount)
            } else {
                displayError()
            }
        }
    }

    private fun signIntoFirebase(googleSignInAccount: GoogleSignInAccount?) {
        firebaseApi().signIntoFirebase(googleSignInAccount, object : AuthenticationCallbacks {
            override fun onSuccess(firebaseUser: FirebaseUser) {
                navigateToThingyAndFinish()
            }

            override fun onAuthenticationFailure() {
                displayError()
            }
        })
    }

    private fun displayError() {
        // TODO: show an error state, ideally with a reason
    }

    companion object {
        private const val REQUEST_CODE_SIGN_IN = 1
    }
}
